File: otc-agent/src/utils/xml-parser.ts
/**
 * Utility to parse XML from agent messages
 */

export interface OTCQuote {
  quoteId: string;
  beneficiary?: string;
  tokenAmount: string;
  tokenAmountFormatted?: string;
  tokenSymbol: string;
  tokenChain?: "ethereum" | "base" | "bsc" | "solana";
  apr?: number;
  lockupMonths: number;
  lockupDays: number;
  pricePerToken?: number;
  priceUsd?: number;
  totalValueUsd?: number;
  totalUsd?: number;
  discountBps: number;
  discountPercent: number;
  discountUsd?: number;
  finalPriceUsd?: number;
  paymentCurrency: string;
  paymentAmount?: string;
  paymentSymbol?: string;
  ethPrice?: number;
  createdAt?: string;
  status?: string;
  message?: string;
  consignmentId?: string;
  isFixedPrice?: boolean;
}

export interface QuoteAccepted {
  quoteId: string;
  offerId: string;
  transactionHash: string;
  tokenAmount: string;
  tokenAmountFormatted: string;
  tokenSymbol: string;
  tokenName: string;
  paidAmount: string;
  paymentCurrency: string;
  discountBps: number;
  discountPercent: number;
  totalSaved: string;
  finalPrice: string;
  status: string;
  timestamp: string;
  message: string;
}

/**
 * Extract XML from message text
 */
export function extractXMLFromMessage(messageText: string): string | null {
  // Try to find XML between comment markers first
  const commentMatch = messageText.match(
    /<!-- XML_START -->([\s\S]*?)<!-- XML_END -->/,
  );
  if (commentMatch && commentMatch[1]) {
    return commentMatch[1].trim();
  }

  // Try to find quote XML (supports lower and PascalCase)
  const quoteMatch = messageText.match(
    /<(quote|Quote)>([\s\S]*?)<\/(quote|Quote)>/,
  );
  if (quoteMatch && quoteMatch[0]) {
    return quoteMatch[0];
  }

  // Try to find quoteAccepted XML (supports lower and PascalCase)
  const acceptedMatch = messageText.match(
    /<(quoteAccepted|QuoteAccepted)>([\s\S]*?)<\/(quoteAccepted|QuoteAccepted)>/,
  );
  if (acceptedMatch && acceptedMatch[0]) {
    return acceptedMatch[0];
  }

  return null;
}

/**
 * Parse quote from XML
 */
export function parseOTCQuoteXML(xmlString: string): OTCQuote | null {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(xmlString, "text/xml");

  // Check for parsing errors
  const parseError = xmlDoc.querySelector("parsererror");
  if (parseError) {
    console.error("XML parsing error:", parseError.textContent);
    return null;
  }

  const getElementText = (tagName: string): string => {
    const elem = xmlDoc.getElementsByTagName(tagName)[0];
    return elem ? elem.textContent || "" : "";
  };

  const getElementNumber = (tagName: string): number => {
    const text = getElementText(tagName);
    return text ? parseFloat(text) : 0;
  };

  // Support both lowercase and PascalCase root tags
  const rootTag =
    xmlDoc.querySelector("Quote") || xmlDoc.querySelector("quote");
  if (!rootTag) {
    console.error("No quote root element found");
    return null;
  }

  const tokenChain = getElementText("tokenChain") || getElementText("chain");

  return {
    quoteId: getElementText("quoteId"),
    tokenAmount: getElementText("tokenAmount"),
    tokenAmountFormatted: getElementText("tokenAmountFormatted"),
    tokenSymbol: getElementText("tokenSymbol"),
    tokenChain: tokenChain
      ? (tokenChain as "ethereum" | "base" | "bsc" | "solana")
      : undefined,
    apr: getElementNumber("apr"),
    lockupMonths: getElementNumber("lockupMonths"),
    lockupDays: getElementNumber("lockupDays"),
    pricePerToken:
      getElementNumber("pricePerToken") || getElementNumber("priceUsdPerToken"),
    totalValueUsd: getElementNumber("totalValueUsd"),
    discountBps: getElementNumber("discountBps"),
    discountPercent: getElementNumber("discountPercent"),
    discountUsd: getElementNumber("discountUsd"),
    finalPriceUsd:
      getElementNumber("finalPriceUsd") || getElementNumber("discountedUsd"),
    paymentCurrency: getElementText("paymentCurrency"),
    paymentAmount: getElementText("paymentAmount"),
    paymentSymbol: getElementText("paymentSymbol"),
    ethPrice: getElementNumber("ethPrice") || undefined,
    createdAt: getElementText("createdAt"),
    status: getElementText("status") || undefined,
    message: getElementText("message"),
  };
}

/**
 * Parse quote accepted XML
 */
export function parseQuoteAcceptedXML(xmlString: string): QuoteAccepted | null {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(xmlString, "text/xml");

  // Check for parsing errors
  const parseError = xmlDoc.querySelector("parsererror");
  if (parseError) {
    console.error("XML parsing error:", parseError.textContent);
    return null;
  }

  const getElementText = (tagName: string): string => {
    const elem = xmlDoc.getElementsByTagName(tagName)[0];
    return elem ? elem.textContent || "" : "";
  };

  const getElementNumber = (tagName: string): number => {
    const text = getElementText(tagName);
    return text ? parseFloat(text) : 0;
  };

  return {
    quoteId: getElementText("quoteId"),
    offerId: getElementText("offerId"),
    transactionHash: getElementText("transactionHash"),
    tokenAmount: getElementText("tokenAmount"),
    tokenAmountFormatted: getElementText("tokenAmountFormatted"),
    tokenSymbol: getElementText("tokenSymbol"),
    tokenName: getElementText("tokenName"),
    paidAmount: getElementText("paidAmount"),
    paymentCurrency: getElementText("paymentCurrency"),
    discountBps: getElementNumber("discountBps"),
    discountPercent: getElementNumber("discountPercent"),
    totalSaved: getElementText("totalSaved"),
    finalPrice: getElementText("finalPrice"),
    status: getElementText("status"),
    timestamp: getElementText("timestamp"),
    message: getElementText("message"),
  };
}

/**
 * Check if message contains a quote
 */
export function messageContainsQuote(messageText: string): boolean {
  return !!(
    messageText.includes("<quote>") ||
    messageText.includes("<quote>") ||
    messageText.includes("<quoteAccepted>") ||
    messageText.includes("<!-- XML_START -->")
  );
}

/**
 * Parse any XML type from message
 */
export function parseMessageXML(messageText: string): {
  type: "otc_quote" | "quote_accepted" | null;
  data: OTCQuote | QuoteAccepted | null;
} {
  const xmlString = extractXMLFromMessage(messageText);

  if (!xmlString) {
    return { type: null, data: null };
  }

  // Try parsing as quote
  if (xmlString.match(/<(quote|Quote)>/)) {
    const quote = parseOTCQuoteXML(xmlString);
    if (quote) {
      return { type: "otc_quote", data: quote };
    }
  }

  // Try parsing as quote accepted
  if (xmlString.match(/<(quoteAccepted|QuoteAccepted)>/)) {
    const accepted = parseQuoteAcceptedXML(xmlString);
    if (accepted) {
      return { type: "quote_accepted", data: accepted };
    }
  }

  return { type: null, data: null };
}

File: cpp_generated/packages/integration/otc_agent/src/utils/xml-parser.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/otc-agent/src/utils/xml-parser.h"

any extractXMLFromMessage(string messageText)
{
    auto commentMatch = messageText->match((new RegExp(std::string("<!-- XML_START -->([\s\S]*?)<!-- XML_END --"))));
    if (AND((commentMatch), ((*const_(commentMatch))[1]))) {
        return (*const_(commentMatch))[1]->trim();
    }
    auto quoteMatch = messageText->match((new RegExp(std::string("<(quote|Quote)>([\s\S]*?)<\/(quote|Quote)"))));
    if (AND((quoteMatch), ((*const_(quoteMatch))[0]))) {
        return (*const_(quoteMatch))[0];
    }
    auto acceptedMatch = messageText->match((new RegExp(std::string("<(quoteAccepted|QuoteAccepted)>([\s\S]*?)<\/(quoteAccepted|QuoteAccepted)"))));
    if (AND((acceptedMatch), ((*const_(acceptedMatch))[0]))) {
        return (*const_(acceptedMatch))[0];
    }
    return nullptr;
};


any parseOTCQuoteXML(string xmlString)
{
    auto parser = std::make_shared<DOMParser>();
    shared xmlDoc = parser->parseFromString(xmlString, std::string("text/xml"));
    auto parseError = xmlDoc->querySelector(std::string("parsererror"));
    if (parseError) {
        console->error(std::string("XML parsing error:"), parseError->textContent);
        return nullptr;
    }
    shared getElementText = [=](auto tagName) mutable
    {
        auto elem = const_(xmlDoc->getElementsByTagName(tagName))[0];
        return (elem) ? any(OR((elem->textContent), (string_empty))) : any(string_empty);
    };
    auto getElementNumber = [=](auto tagName) mutable
    {
        auto text = getElementText(tagName);
        return (text) ? any(parseFloat(text)) : any(0);
    };
    auto rootTag = OR((xmlDoc->querySelector(std::string("Quote"))), (xmlDoc->querySelector(std::string("quote"))));
    if (!rootTag) {
        console->error(std::string("No quote root element found"));
        return nullptr;
    }
    auto tokenChain = OR((getElementText(std::string("tokenChain"))), (getElementText(std::string("chain"))));
    return object{
        object::pair{std::string("quoteId"), getElementText(std::string("quoteId"))}, 
        object::pair{std::string("tokenAmount"), getElementText(std::string("tokenAmount"))}, 
        object::pair{std::string("tokenAmountFormatted"), getElementText(std::string("tokenAmountFormatted"))}, 
        object::pair{std::string("tokenSymbol"), getElementText(std::string("tokenSymbol"))}, 
        object::pair{std::string("tokenChain"), (tokenChain) ? any((as<any>(tokenChain))) : any(undefined)}, 
        object::pair{std::string("apr"), getElementNumber(std::string("apr"))}, 
        object::pair{std::string("lockupMonths"), getElementNumber(std::string("lockupMonths"))}, 
        object::pair{std::string("lockupDays"), getElementNumber(std::string("lockupDays"))}, 
        object::pair{std::string("pricePerToken"), OR((getElementNumber(std::string("pricePerToken"))), (getElementNumber(std::string("priceUsdPerToken"))))}, 
        object::pair{std::string("totalValueUsd"), getElementNumber(std::string("totalValueUsd"))}, 
        object::pair{std::string("discountBps"), getElementNumber(std::string("discountBps"))}, 
        object::pair{std::string("discountPercent"), getElementNumber(std::string("discountPercent"))}, 
        object::pair{std::string("discountUsd"), getElementNumber(std::string("discountUsd"))}, 
        object::pair{std::string("finalPriceUsd"), OR((getElementNumber(std::string("finalPriceUsd"))), (getElementNumber(std::string("discountedUsd"))))}, 
        object::pair{std::string("paymentCurrency"), getElementText(std::string("paymentCurrency"))}, 
        object::pair{std::string("paymentAmount"), getElementText(std::string("paymentAmount"))}, 
        object::pair{std::string("paymentSymbol"), getElementText(std::string("paymentSymbol"))}, 
        object::pair{std::string("ethPrice"), OR((getElementNumber(std::string("ethPrice"))), (undefined))}, 
        object::pair{std::string("createdAt"), getElementText(std::string("createdAt"))}, 
        object::pair{std::string("status"), OR((getElementText(std::string("status"))), (undefined))}, 
        object::pair{std::string("message"), getElementText(std::string("message"))}
    };
};


any parseQuoteAcceptedXML(string xmlString)
{
    auto parser = std::make_shared<DOMParser>();
    shared xmlDoc = parser->parseFromString(xmlString, std::string("text/xml"));
    auto parseError = xmlDoc->querySelector(std::string("parsererror"));
    if (parseError) {
        console->error(std::string("XML parsing error:"), parseError->textContent);
        return nullptr;
    }
    shared getElementText = [=](auto tagName) mutable
    {
        auto elem = const_(xmlDoc->getElementsByTagName(tagName))[0];
        return (elem) ? any(OR((elem->textContent), (string_empty))) : any(string_empty);
    };
    auto getElementNumber = [=](auto tagName) mutable
    {
        auto text = getElementText(tagName);
        return (text) ? any(parseFloat(text)) : any(0);
    };
    return object{
        object::pair{std::string("quoteId"), getElementText(std::string("quoteId"))}, 
        object::pair{std::string("offerId"), getElementText(std::string("offerId"))}, 
        object::pair{std::string("transactionHash"), getElementText(std::string("transactionHash"))}, 
        object::pair{std::string("tokenAmount"), getElementText(std::string("tokenAmount"))}, 
        object::pair{std::string("tokenAmountFormatted"), getElementText(std::string("tokenAmountFormatted"))}, 
        object::pair{std::string("tokenSymbol"), getElementText(std::string("tokenSymbol"))}, 
        object::pair{std::string("tokenName"), getElementText(std::string("tokenName"))}, 
        object::pair{std::string("paidAmount"), getElementText(std::string("paidAmount"))}, 
        object::pair{std::string("paymentCurrency"), getElementText(std::string("paymentCurrency"))}, 
        object::pair{std::string("discountBps"), getElementNumber(std::string("discountBps"))}, 
        object::pair{std::string("discountPercent"), getElementNumber(std::string("discountPercent"))}, 
        object::pair{std::string("totalSaved"), getElementText(std::string("totalSaved"))}, 
        object::pair{std::string("finalPrice"), getElementText(std::string("finalPrice"))}, 
        object::pair{std::string("status"), getElementText(std::string("status"))}, 
        object::pair{std::string("timestamp"), getElementText(std::string("timestamp"))}, 
        object::pair{std::string("message"), getElementText(std::string("message"))}
    };
};


boolean messageContainsQuote(string messageText)
{
    return !!(OR((OR((OR((messageText->includes(std::string("<quote>"))), (messageText->includes(std::string("<quote>"))))), (messageText->includes(std::string("<quoteAccepted>"))))), (messageText->includes(std::string("<!-- XML_START -->")))));
};


object parseMessageXML(string messageText)
{
    auto xmlString = extractXMLFromMessage(messageText);
    if (!xmlString) {
        return object{
            object::pair{std::string("type"), nullptr}, 
            object::pair{std::string("data"), nullptr}
        };
    }
    if (xmlString->match((new RegExp(std::string("<(quote|Quote)"))))) {
        auto quote = parseOTCQuoteXML(xmlString);
        if (quote) {
            return object{
                object::pair{std::string("type"), std::string("otc_quote")}, 
                object::pair{std::string("data"), quote}
            };
        }
    }
    if (xmlString->match((new RegExp(std::string("<(quoteAccepted|QuoteAccepted)"))))) {
        auto accepted = parseQuoteAcceptedXML(xmlString);
        if (accepted) {
            return object{
                object::pair{std::string("type"), std::string("quote_accepted")}, 
                object::pair{std::string("data"), accepted}
            };
        }
    }
    return object{
        object::pair{std::string("type"), nullptr}, 
        object::pair{std::string("data"), nullptr}
    };
};


